File: BlockchainFormation/blockchain_specifics/LevelDB/LevelDB.py
#  Copyright 2020 ChainLab
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.


import os
import time
import numpy as np
from BlockchainFormation.utils.utils import *
import sys

import hashlib


def leveldb_shutdown(config, logger, ssh_clients, scp_clients):
    """
    runs the LevelDB specific shutdown operations (e.g. pulling the associated logs from the VMs)
    :return:
    """

    pass

def leveldb_startup(config, logger, ssh_clients, scp_clients):
    """
    Runs the LevelDB specific startup script
    :return:
    """

    dir_name = os.path.dirname(os.path.realpath(__file__))

    # the indices of the blockchain nodes
    config['node_indices'] = list(range(0, config['vm_count']))

    # Creating docker swarm
    logger.info("Preparing & starting docker swarm")

    stdin, stdout, stderr = ssh_clients[0].exec_command("sudo docker swarm init")
    out = stdout.readlines()
    # for index, _ in enumerate(out):
    #     logger.debug(out[index].replace("\n", ""))

    # logger.debug("".join(stderr.readlines()))

    stdin, stdout, stderr = ssh_clients[0].exec_command("sudo docker swarm join-token manager")
    out = stdout.readlines()
    # logger.debug(out)
    # logger.debug("".join(stderr.readlines()))
    join_command = out[2].replace("    ", "").replace("\n", "")

    for index, _ in enumerate(config['priv_ips']):

        if index != 0:
            stdin, stdout, stderr = ssh_clients[index].exec_command("sudo " + join_command)
            out = stdout.readlines()
            # logger.debug(out)
            # logger.debug("".join(stderr.readlines()))

    config['join_command'] = "sudo " + join_command
    # Name of the swarm network
    my_net = "my-net"
    stdin, stdout, stderr = ssh_clients[0].exec_command(f"sudo docker network create --subnet 10.10.0.0/16 --attachable --driver overlay {my_net}")
    out = stdout.readlines()
    # logger.debug(out)
    # logger.debug("".join(stderr.readlines()))
    network = out[0].replace("\n", "")

    logger.info("Testing whether setup was successful")
    stdin, stdout, stderr = ssh_clients[0].exec_command("sudo docker node ls")
    out = stdout.readlines()
    for index, _ in enumerate(out):
        logger.debug(out[index].replace("\n", ""))

    logger.debug("".join(stderr.readlines()))
    if len(out) == len(config['priv_ips']) + 1:
        logger.info("Docker swarm started successfully")
    else:
        logger.info("Docker swarm setup was not successful")
        sys.exit("Fatal error when performing docker swarm setup")

    scp_clients[0].put(f"{dir_name}/setup", "/home/ubuntu", recursive=True)

    channel = ssh_clients[0].get_transport().open_session()
    channel.exec_command(f"(cd setup && . ~/.profile && npm install >> ~/install.log && echo 'Success' >> ~/install.log)")
    wait_till_done(config, ssh_clients, config['ips'], 180, 10, "~/install.log", "Success", 30, logger)


    logger.info("Starting the server")

    stdin, stdout, stderr = ssh_clients[0].exec_command("echo '{\n    \"ip\": \"" + f"{config['priv_ips'][0]}" + "\"\n}' >> ~/setup/config.json")
    logger.debug(stdout.readlines())
    logger.debug(stderr.readlines())

    channel = ssh_clients[0].get_transport().open_session()
    channel.exec_command(f"(cd setup && node server.js)")




def leveldb_restart(config, logger, ssh_clients, scp_clients):

    pass

File: BlockchainFormation/blockchain_specifics/leveldb/setup/leveldb.js
/*
 * Copyright 2019  ChainLab
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const level = require('level');
const db = level('my-db');
// 1) Create our database, supply location and options.
//    This will create or open the underlying store.

class levelDB {

  async init() {
    return Promise.resolve(1);
  }

  async writeData(key, value) {

    await db.put(key, value).catch(err => {
      console.log("An error occurred");
      return Promise.reject("Error");
    });
    console.log("Wrote key " + key);
    return Promise.resolve();
  }

  async readData(key) {

    let value = await db.get(key).catch(err => {
      console.log("An error occurred");
      return Promise.reject("Error");
    });
    console.log("Read value " + value);
    return Promise.resolve(value);
  }

  async matrixMultiplication(n) {
        // console.log(n)
    async function multiplyMatrices(m1, m2) {
        var result = [];
        for (var i = 0; i < m1.length; i++) {
            result[i] = [];
            for (var j = 0; j < m2[0].length; j++) {
                var sum = 0;
                for (var k = 0; k < m1[0].length; k++) {
                    sum += m1[i][k] * m2[k][j];
                }
                result[i][j] = sum;
            }
        }
        return result;
    }

    var f = 1;
    var m1 = [];
    for (var i = 0; i < n; i++) {
        //console.log(i);
        m1[i] = [];
        for (var j = 0; j < n; j++) {
            m1[i][j] = f;
            f++;
        }
    }

    var m2 = m1;

    var result = await multiplyMatrices(m1, m2).catch(err => console.log(err()));

    var matrixSum = 0;

    for (var i = 0; i < result.length; ++i) {
        for (var j = 0; j < result[i].length; ++j) {
            matrixSum += result[i][j];
        }
    }
    return Promise.resolve(matrixSum.toString());
  }


  async close() {
    return Promise.resolve(1);
  }
}

module.exports = levelDB;
